### PR TITLE
Adds test for synapse replay with virtual population

### LINKED ISSUE #96: Test using projections
Users create projections which instantiate synapses on neurons while not having a presynaptic cell on any rank

These can be activated by the replay stimulus feature

### PR DESCRIPTION
**Pull request by juanjosegarcan**
## Context
Add a unit test using projections

Fixes #96 
## Scope
Based on the ring test simulation circuit test a replay synapse input changing RingA population type to virtual, testing that way the use of projections

## Testing
Extended and improved test_replay_manager 

## Review
* [ ] PR description is complete
* [ ] Coding style (imports, function length, New functions, classes or files) are good
* [ ] Unit/Scientific test added
* [ ] Updated Readme, in-code, developer documentation

### CODE REVIEW

File: tests/unit/test_replay_manager.py [hunk 1]
```diff
@@ -1,94 +1,261 @@
...
+    nd.run()
+
+    v_increase_rate = np.diff(voltage_vec, 2)
+    window_sum = np.convolve(v_increase_rate, [1, 2, 4, 2, 1], 'valid')
+    strong_reduction_pos = np.nonzero(window_sum < -0.5)[0]
+    non_consecutives_pos = strong_reduction_pos[np.insert(
+        np.diff(strong_reduction_pos) > 1, 0, True)]
+    expected_positions = np.array([32, 75])
+
+    npt.assert_equal(non_consecutives_pos, expected_positions)
```
Comment (cattabiani) [inline tests/unit/test_replay_manager.py:198]: what about putting this in `tests/utils.py`? It is used in various tests. I also have a PR about it if you want to wait

Comment (cattabiani) [inline tests/unit/test_replay_manager.py:198]: in addition: why didn't you opt for detecting the peaks in the trace directly using something like `scipy.signal.find_peaks`? Why the second derivative? Why the [1, 2, 4, 2, 1] filtering kernel? 

these are all curiosities that came to me later, after your previous PR

File: tests/unit/test_replay_manager.py [hunk 2]
```diff
@@ -1,94 +1,261 @@
...
```
Comment (cattabiani) [inline tests/unit/test_replay_manager.py:261]: the same as before: this may go in a function

**Review by cattabiani — COMMENTED** [review]
very nice! a few suggestions

File: tests/unit/test_replay_manager.py (hunk 1 above)
Comment (cattabiani) [reply tests/unit/test_replay_manager.py:198]: I also see [here](https://nrn.readthedocs.io/en/8.2.6/tutorials/ball-and-stick-4.html) that neuron has already a spike detector. Maybe we can use that

Comment (cattabiani) [reply tests/unit/test_replay_manager.py:198]: apparently it works only if you have netcons attached to the cell and it is a rudimentary threshold crossing detector. Maybe it is better to have this function instead.

Comment (cattabiani) [reply tests/unit/test_replay_manager.py:198]: another small thing: the filtering kernel does not have an integral == 1. In fact, this one is equal to 10. This changes the threshold detection (-0.5 I think). We should have a kernel that has an integral equal to 1 so we can swap them without the need to change the thresholds

**Comment by juanjosegarcan** [comment]
@cattabiani I added a check_signal_peaks function based on scipy.signal.find_peaks to tests utils following your suggestions. Check it please and let me know what you think. May we should use it all around the tests?

File: tests/unit/test_replay_manager.py [hunk 3]
```diff
@@ -1,94 +1,249 @@
...
+
+    # get voltage variations in cell 1001
+    cell = nd.circuits.get_node_manager("RingB").get_cell(1001)
+    voltage_vec = Nd.Vector()
+    voltage_vec.record(cell._cellref.soma[0](0.5)._ref_v)
+
+    Nd.finitialize()
+    nd.run()
+
+    expected_positions = np.array([39, 81])
```
Comment (cattabiani) [inline tests/unit/test_replay_manager.py:248]: are you sure you need to convert it to a `np.array` explicitly? I think it works with a simple list too

File: tests/unit/test_test_utils.py [hunk 4]
```diff
@@ -114,3 +115,9 @@ def test_check_synapses(create_tmp_simulation_config_file):
     # wrong type
     with pytest.raises(TypeError):
         utils.check_synapses([0], edges, selection)
+
+
+def test_check_signal_peaks():
```
Comment (cattabiani) [inline tests/unit/test_test_utils.py:120]: I am not sure that this function works even if the signal rises or descends. May you test also that case? An inclined trace with some peaks

File: tests/utils.py [hunk 5]
```diff
@@ -186,3 +187,26 @@ def check_synapse(syn, edges, selection, **kwargs):
...
+        taken as reference.
+        threshold: peak detection threshold with values
+        between 0 and 1, measured over the normalized
+        signal.
+
+    Raises:
+        AssertionError: If any of the reference peak
+        positions doesn't match with the obtained peaks
+    """
+    x_norm = (x - x.min()) / (x.max() - x.min())
```
Comment (cattabiani) [inline tests/utils.py:209]: using prominence is usually good but I am concerned about this norm. May you try the test where you combine a spiking signal with a steady rising one? 

y = spike_signal + a * t

where a is a fixed coefficient. It may happen if we are measuring accumulations of ions or even voltage or current

**Review by cattabiani — COMMENTED** [review]
I am concerned about an edge case. I would like a test for that case too

File: tests/unit/test_replay_manager.py (hunk 3 above)
Comment (cattabiani) [reply tests/unit/test_replay_manager.py:248]: even in that case you can handle the conversion in the function itself no?

Comment (cattabiani) [reply tests/unit/test_replay_manager.py]: this is still missing no?

File: tests/utils.py [hunk 6]
```diff
@@ -237,3 +238,25 @@ def check_synapse(syn, edges, selection, **kwargs):
...
+        respect of the surrounding baseline of the signal
+
+    Raises:
+        AssertionError: If any of the reference peak
+        positions doesn't match with the obtained peaks
+    """
+    x = np.array(x)
+    peaks_pos = find_peaks(x, prominence=threshold)[0]
+
+    np.testing.assert_equal(peaks_pos, ref_peaks_pos)
```
Comment (cattabiani) [inline tests/utils.py:262]:
```suggestion
    np.testing.assert_equal(peaks_pos, np.array(ref_peaks_pos))
```

**Review by cattabiani — COMMENTED** [review]
I think 1 change and we are done

File: tests/utils.py (hunk 6 above)
Comment (juanjosegarcan) [reply tests/utils.py:262]: that's not really needed right?

Comment (cattabiani) [reply tests/utils.py:260]: I did not see the x = np.array(x) before

Comment (juanjosegarcan) [reply tests/utils.py:260]: I removed too

File: tests/unit/test_test_utils.py (hunk 4 above)
Comment (cattabiani) [reply tests/unit/test_test_utils.py:191]: done

**Review by cattabiani — APPROVED** [review]
LGTM!
